pkg/handlers/event_stream_test.go

```diff
@@ -0,0 +1,212 @@
+	"github.com/optimizely/agent/pkg/optimizely"
+	"github.com/optimizely/agent/pkg/optimizely/optimizelytest"
+
+	"github.com/go-chi/chi"
+	"github.com/optimizely/go-sdk/pkg/entities"
+	"github.com/stretchr/testify/suite"
+)
+
+type EventStreamTestSuite struct {
+	suite.Suite
+	tc  *optimizelytest.TestClient
+	mux *chi.Mux
+}
+
+type EventStreamMW struct {
+	optlyClient *optimizely.OptlyClient
+}
+
+func (o *EventStreamMW) ClientCtx(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := context.WithValue(r.Context(), middleware.OptlyClientKey, o.optlyClient)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+func (o *EventStreamMW) UserCtx(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		optlyContext := optimizely.NewContext("testUser", make(map[string]interface{}))
+		ctx := context.WithValue(r.Context(), middleware.OptlyContextKey, optlyContext)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// Setup Mux
+func (suite *EventStreamTestSuite) SetupTest() {
+	testClient := optimizelytest.NewClient()
+	optlyClient := &optimizely.OptlyClient{
+		OptimizelyClient: testClient.OptimizelyClient,
+		ConfigManager:    nil,
+		ForcedVariations: testClient.ForcedVariations,
+	}
+
+	mux := chi.NewMux()
+	eventsAPI := NewEventStreamHandler()
+	EventStreamMW := &EventStreamMW{optlyClient}
+
+	mux.Use(EventStreamMW.ClientCtx, EventStreamMW.UserCtx)
+	mux.Get("/notifications/event-stream", eventsAPI.HandleEventSteam)
+
+	suite.mux = mux
+	suite.tc = testClient
+}
+
+func (suite *EventStreamTestSuite) TestFeatureTestStream() {
+	feature := entities.Feature{Key: "one"}
+	suite.tc.AddFeatureTest(feature)
+
+	req := httptest.NewRequest("GET", "/notifications/event-stream", nil)
+	rec := httptest.NewRecorder()
+
+	go func() {
+		// sleep for 100 milliseconds to allow the handler to start
+		time.Sleep(100)
+		suite.tc.OptimizelyClient.IsFeatureEnabled("one", entities.UserContext{"testUser", make(map[string]interface{})})
+	}()
+
+	// create a cancelable request context
+	ctx := req.Context()
+	ctx1,cancelFun := context.WithCancel(ctx)
+	// wait 1 second for the request to be serviced and then cancel the request which closes the request
+	// and notifies the handler. This causes the handler to shuts down properly
+	go func() {
+		time.Sleep(1 * time.Second)
+		cancelFun()
+	}()
+
+	suite.mux.ServeHTTP(rec, req.WithContext(ctx1))
+
+	expected := "data: {\"Type\":\"feature\",\"UserContext\":{\"ID\":\"testUser\",\"Attributes\":{}},\"DecisionInfo\":{\"feature\":{\"featureEnabled\":true,\"featureKey\":\"one\",\"source\":\"feature-test\",\"sourceInfo\":{\"experimentKey\":\"1\",\"variationKey\":\"2\"}}}}\n\n"
+
+	suite.Equal(http.StatusOK, rec.Code)
+
+	// Unmarshal response
+	response := string(rec.Body.Bytes())
+	suite.Equal(expected,response)
+}
+
+func (suite *EventStreamTestSuite) TestActivateExperiment() {
```

Comment:
I think if we invert the goroutine and use a `sync.WaitGroup` then the tests will be a bit easier to manage and we don't need to rely on `sleep` statements which are inherently flaky. 
For example:
```go
func (suite *EventStreamTestSuite) TestActivateExperiment() {
	testVariation := suite.tc.ProjectConfig.CreateVariation("variation_a")
	suite.tc.AddExperiment("one", []entities.Variation{testVariation})

	req := httptest.NewRequest("GET", "/notifications/event-stream", nil)
	rec := httptest.NewRecorder()
	// create a cancelable request context
	ctx, cancelFun := context.WithCancel(req.Context())
	expected := `data: {"Type":"ab-test","UserContext":{"ID":"testUser","Attributes":{}},"DecisionInfo":{"experimentKey":"one","variationKey":"variation_a"}}` + "\n\n"

	wg := sync.WaitGroup{}
	wg.Add(1)

	go func() {
		suite.mux.ServeHTTP(rec, req.WithContext(ctx))
		suite.Equal(http.StatusOK, rec.Code)

		// Unmarshal response
		response := string(rec.Body.Bytes())
		suite.Equal(expected, response)
		wg.Done()
	}()

	suite.tc.OptimizelyClient.Activate("one", entities.UserContext{"testUser", make(map[string]interface{})})
	cancelFun()

	wg.Wait()
}
```

Similar comment on all the tests relying on `time.Sleep` in this suite.

Reply:
this might require another WaitGroup to be super certain the request was initiated, but the concept is sound.